cmd/maestro/server/grpc_server.go

```diff
@@ -257,18 +293,46 @@ func (svr *GRPCServer) Subscribe(subReq *pbv1.SubscriptionRequest, subServer pbv
 			return fmt.Errorf("failed to convert cloudevent to protobuf: %v", err)
 		}
 
-		// send the cloudevent to the subscriber
-		// TODO: error handling to address errors beyond network issues.
-		if err := subServer.Send(pbEvt); err != nil {
-			return err
+		select {
+		case sendCh <- pbEvt:
+		case <-ctx.Done():
+			return status.Error(codes.Unavailable, "stream context canceled")
 		}
 
 		return nil
 	})
 
+	go func() {
+		ticker := time.NewTicker(svr.heartbeatCheckInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				heartbeat := &pbv1.CloudEvent{
+					SpecVersion: "1.0",
+					Id:          uuid.New().String(),
+					Type:        types.HeartbeatCloudEventsType,
+				}
+
+				select {
+				case sendCh <- heartbeat:
+				default:
+					log.Warn("send channel is full, dropping heartbeat")
```

Comment:
is it safe to drop heartbeat event?

Reply:
agree, it seems a busy channel will cause drop of the heartbeat. And if the heartbeat is dropped, it will not retry but wait for the next interval.

Reply:
shouldn't we separate heartbeat and normal event channel?